diff --git a/Makefile.emscripten b/Makefile.emscripten
index fb74177eed..2221d5ca6f 100644
--- a/Makefile.emscripten
+++ b/Makefile.emscripten
@@ -121,8 +121,11 @@ PTHREAD_POOL_SIZE ?= 4
 ASYNCIFY_ADD ?= dynCall_*,emscripten_mainloop
 ASYNCIFY_REMOVE ?= threaded_worker
 
+EXIT_RUNTIME ?= 1
+
 STACK_SIZE ?= 4194304
 INITIAL_HEAP ?= 134217728
+ASYNCIFY_STACK_SIZE ?= 8192
 
 # 4194304 ----- 4 MiB (Stack: recommended)
 # 8388608 ----- 8 MiB
@@ -142,6 +145,9 @@ EXPORTED_FUNCTIONS = _main,_malloc,_free,_cmd_savefiles,_cmd_save_state,_cmd_loa
 _cmd_toggle_menu,_cmd_reload_config,_cmd_toggle_grab_mouse,_cmd_toggle_game_focus,_cmd_reset,_cmd_toggle_pause,_cmd_pause,_cmd_unpause,\
 _cmd_set_volume,_cmd_set_shader,_cmd_cheat_set_code,_cmd_cheat_get_code,_cmd_cheat_toggle_index,_cmd_cheat_get_code_state,_cmd_cheat_realloc,\
 _cmd_cheat_get_size,_cmd_cheat_apply_cheats,EmscriptenSendCommand,EmscriptenReceiveCommandReply
+EM_NULL :=
+EM_SPACE := $(EM_NULL) $(EM_NULL)
+EXPORTED_FUNCTIONS_NO_SPACE := $(subst $(EM_SPACE),,$(EXPORTED_FUNCTIONS))
 
 EXPORTS := callMain,FS,PATH,ERRNO_CODES,ENV,stringToNewUTF8,UTF8ToString,Browser,EmscriptenSendCommand,EmscriptenReceiveCommandReply
 
@@ -182,9 +188,9 @@ ifeq ($(HAVE_SDL2), 1)
    DEFINES += -DHAVE_SDL2
 endif
 
-LDFLAGS := -L. --no-heap-copy -s STACK_SIZE=$(STACK_SIZE) -s INITIAL_MEMORY=$(INITIAL_HEAP) \
-           -s EXPORTED_RUNTIME_METHODS=$(EXPORTS) -s EXIT_RUNTIME=1 \
-           -s ALLOW_MEMORY_GROWTH=1 -s EXPORTED_FUNCTIONS="$(EXPORTED_FUNCTIONS)" \
+LDFLAGS := -L. --no-heap-copy -s TOTAL_STACK=$(STACK_SIZE) -s TOTAL_MEMORY=$(INITIAL_HEAP) \
+           -s EXPORTED_RUNTIME_METHODS=[$(EXPORTS)] -s EXIT_RUNTIME=$(EXIT_RUNTIME) \
+           -s ALLOW_MEMORY_GROWTH=1 -s EXPORTED_FUNCTIONS=[$(EXPORTED_FUNCTIONS_NO_SPACE)] \
            -s MODULARIZE=1 -s EXPORT_ES6=1 -s EXPORT_NAME="libretro_$(subst -,_,$(LIBRETRO))" \
            -s DISABLE_DEPRECATED_FIND_EVENT_TARGET_BEHAVIOR=0 \
            -s ENVIRONMENT=web,worker -s WASM_BIGINT=1 \
@@ -272,13 +278,13 @@ ifeq ($(JSPI), 1)
    LDFLAGS += -s JSPI=1 -fwasm-exceptions
 else ifeq ($(MIN_ASYNC), 1)
    DEFINES += -DEMSCRIPTEN_ASYNCIFY -DEMSCRIPTEN_MIN_ASYNCIFY
-   LDFLAGS += -s ASYNCIFY=1 -s ASYNCIFY_STACK_SIZE=8192 -s ASYNCIFY_IGNORE_INDIRECT=1 -s ASYNCIFY_ADD='$(ASYNCIFY_ADD)' -s ASYNCIFY_REMOVE='$(ASYNCIFY_REMOVE)'
+   LDFLAGS += -s ASYNCIFY=1 -s ASYNCIFY_STACK_SIZE=$(ASYNCIFY_STACK_SIZE) -s ASYNCIFY_IGNORE_INDIRECT=1 -s ASYNCIFY_ADD='$(ASYNCIFY_ADD)' -s ASYNCIFY_REMOVE='$(ASYNCIFY_REMOVE)'
    ifeq ($(DEBUG), 1)
       LDFLAGS += -s ASYNCIFY_ADVISE #-s ASYNCIFY_DEBUG=1
    endif
 else ifeq ($(ASYNC), 1)
    DEFINES += -DEMSCRIPTEN_ASYNCIFY -DEMSCRIPTEN_FULL_ASYNCIFY
-   LDFLAGS += -s ASYNCIFY=1 -s ASYNCIFY_STACK_SIZE=8192
+   LDFLAGS += -s ASYNCIFY=1 -s ASYNCIFY_STACK_SIZE=$(ASYNCIFY_STACK_SIZE)
    ifeq ($(DEBUG), 1)
       #LDFLAGS += -s ASYNCIFY_DEBUG=1 # broken?
    endif
diff --git a/emscripten/library_platform_emscripten.js b/emscripten/library_platform_emscripten.js
index 79c69fd986..e92a9d38f7 100644
--- a/emscripten/library_platform_emscripten.js
+++ b/emscripten/library_platform_emscripten.js
@@ -15,7 +15,7 @@ var LibraryPlatformEmscripten = {
 
    PlatformEmscriptenKeepThreadAlive: function() {
       // keeps the thread alive until exited explicitly
-      {{{ runtimeKeepalivePush() }}}
+      noExitRuntime = true;
    },
 
    $PlatformEmscriptenOnWindowResize: function() {
@@ -260,4 +260,4 @@ var LibraryPlatformEmscripten = {
 };
 
 autoAddDeps(LibraryPlatformEmscripten, '$RPE');
-addToLibrary(LibraryPlatformEmscripten);
+mergeInto(LibraryManager.library, LibraryPlatformEmscripten);
diff --git a/emscripten/library_rwebaudio.js b/emscripten/library_rwebaudio.js
index d97bfead0b..f90459c79a 100644
--- a/emscripten/library_rwebaudio.js
+++ b/emscripten/library_rwebaudio.js
@@ -112,4 +112,4 @@ var LibraryRWebAudio = {
 };
 
 autoAddDeps(LibraryRWebAudio, '$RWA');
-addToLibrary(LibraryRWebAudio);
+mergeInto(LibraryManager.library, LibraryRWebAudio);
diff --git a/emscripten/library_rwebcam.js b/emscripten/library_rwebcam.js
index 44bc8ab333..b9994526a4 100644
--- a/emscripten/library_rwebcam.js
+++ b/emscripten/library_rwebcam.js
@@ -282,4 +282,4 @@ var LibraryRWebCam = {
 };
 
 autoAddDeps(LibraryRWebCam, '$RWC');
-addToLibrary(LibraryRWebCam);
+mergeInto(LibraryManager.library, LibraryRWebCam);
diff --git a/input/drivers_joypad/rwebpad_joypad.c b/input/drivers_joypad/rwebpad_joypad.c
index f215d66b24..3d53edb008 100644
--- a/input/drivers_joypad/rwebpad_joypad.c
+++ b/input/drivers_joypad/rwebpad_joypad.c
@@ -18,6 +18,7 @@
 
 #include <boolean.h>
 #include <retro_miscellaneous.h>
+#include <emscripten.h>
 #include <emscripten/html5.h>
 #include "../../frontend/drivers/platform_emscripten.h"
 
diff --git a/retroarch.c b/retroarch.c
index b4b8a1da0d..21ebc74e61 100644
--- a/retroarch.c
+++ b/retroarch.c
@@ -7062,7 +7062,6 @@ static void retroarch_print_help(const char *arg0)
 #endif
 }
 
-#ifdef HAVE_DYNAMIC
 static void retroarch_parse_input_libretro_path(
       struct rarch_state *p_rarch,
       const char *path, size_t path_len)
@@ -7182,6 +7181,14 @@ static void retroarch_parse_input_libretro_path(
          runloop_set_current_core_type(CORE_TYPE_VIDEO_PROCESSOR, true);
          return;
       }
+      else if (
+               !memcmp(path, "mupen64plus_next", STRLEN_CONST("mupen64plus_next")) ||
+               !memcmp(path, "mupen64plus_libretro_nx", STRLEN_CONST("mupen64plus_libretro_nx")))
+      {
+         /* Embedded static core via libretro_emscripten.bc */
+         runloop_set_current_core_type(CORE_TYPE_PLAIN, true);
+         return;
+      }
       command_event(CMD_EVENT_CORE_INFO_INIT, NULL);
       _len = strlcpy(tmp_path, path, sizeof(tmp_path));
       if (!string_ends_with_size(tmp_path, "_libretro",
@@ -7215,8 +7222,6 @@ end:
             " or directory. Ignoring.\n",
             path ? path : "");
 }
-#endif
-
 #if defined(HAVE_LIBRETRODB) && defined(HAVE_MENU)
 void handle_dbscan_finished(retro_task_t *task,
       void *task_data, void *user_data, const char *err);
@@ -7257,9 +7262,7 @@ static bool retroarch_parse_input_and_config(
 #endif
 
    const struct option opts[]      = {
-#ifdef HAVE_DYNAMIC
       { "libretro",           1, NULL, 'L' },
-#endif
       { "menu",               0, NULL, RA_OPT_MENU },
       { "help",               0, NULL, 'h' },
       { "save",               1, NULL, 's' },
@@ -7366,7 +7369,11 @@ static bool retroarch_parse_input_and_config(
     */
 
    if (!(runloop_st->flags & RUNLOOP_FLAG_HAS_SET_CORE))
+#if defined(EMSCRIPTEN) && !defined(HAVE_STATIC_DUMMY)
+      runloop_set_current_core_type(CORE_TYPE_PLAIN, false);
+#else
       runloop_set_current_core_type(CORE_TYPE_DUMMY, false);
+#endif
 
    path_clear(RARCH_PATH_SUBSYSTEM);
 
@@ -7388,8 +7395,8 @@ static bool retroarch_parse_input_and_config(
 
    /* Make sure we can call retroarch_parse_input several times ... */
    optind                          = 0;
-   optstring                       = "hs:fvVS:A:U:DN:d:e:"
-      BSV_MOVIE_ARG NETPLAY_ARG DYNAMIC_ARG FFMPEG_RECORD_ARG CONFIG_FILE_ARG;
+   optstring                       = "hs:fvVS:A:U:DN:d:e:L:"
+      BSV_MOVIE_ARG NETPLAY_ARG FFMPEG_RECORD_ARG CONFIG_FILE_ARG;
 
 #if defined(WEBOS)
    if (argc > 1 && argv[1][0] == '{')
@@ -7668,11 +7675,9 @@ static bool retroarch_parse_input_and_config(
 #endif
                break;
 
-#ifdef HAVE_DYNAMIC
             case 'L':
                retroarch_parse_input_libretro_path(p_rarch, optarg, strlen(optarg));
                break;
-#endif
             case 'P':
 #ifdef HAVE_BSV_MOVIE
                {
